In [10]:
import numpy as np
np.set_printoptions(precision=2, suppress=True)

1. Найти с помощью NumPy SVD для матрицы

$$\begin{pmatrix}
1 & 2 & 0\\ 
0 & 0 & 5\\ 
3 & -4 & 2\\ 
1 & 6 & 5\\ 
0 & 1 & 0
\end{pmatrix}.$$

In [11]:
a = np.array([[1,2,0],[0,0,5], [3,-4,2], [1,6,5], [0,1,0]])

In [12]:
U, s, W = np.linalg.svd(a)

In [13]:
U

array([[ 0.17,  0.16, -0.53, -0.8 , -0.16],
       [ 0.39, -0.53,  0.61, -0.43,  0.03],
       [-0.14, -0.82, -0.52,  0.14,  0.07],
       [ 0.89,  0.06, -0.25,  0.38, -0.06],
       [ 0.08,  0.11, -0.08, -0.11,  0.98]])

In [14]:
s

array([8.82, 6.14, 2.53])

In [15]:
V = W.T
V

array([[ 0.07, -0.37, -0.93],
       [ 0.72,  0.67, -0.21],
       [ 0.69, -0.65,  0.31]])

In [16]:
D = np.zeros_like(a, dtype=float)
D[np.diag_indices(min(a.shape))] = s
D

array([[8.82, 0.  , 0.  ],
       [0.  , 6.14, 0.  ],
       [0.  , 0.  , 2.53],
       [0.  , 0.  , 0.  ],
       [0.  , 0.  , 0.  ]])

In [17]:
np.dot(np.dot(U,D), V.T)

array([[ 1.,  2.,  0.],
       [ 0., -0.,  5.],
       [ 3., -4.,  2.],
       [ 1.,  6.,  5.],
       [-0.,  1.,  0.]])

2. Для матрицы из предыдущего задания найти:

а) евклидову норму;

$$\left \| A \right \|_{E}=8.82.$$

б) норму Фробениуса.

$$\left \| A \right \|_{F}=\sqrt{\sum_{k=1}^{r}\mu_{k}^{2}}.$$

$$\left \| A \right \|_{F}=\sqrt{8.82^2 + 6.14^2 + 2.53^2} = 11.04$$